crates/codegen/parser/runtime/src/query_engine_tests.rs

```diff
@@ -0,0 +1,169 @@
+use std::collections::{BTreeMap, HashMap};
+
+use super::cst::{NamedNode, Node};
+use super::cursor::Cursor;
+use super::kinds::{RuleKind, TokenKind};
+use super::query_model::Query;
+use super::text_index::TextIndex;
+
+fn token(kind: TokenKind, text: &str) -> NamedNode {
+    NamedNode {
+        name: None,
+        node: Node::token(kind, text.to_string()),
+    }
+}
+
+fn rule<const N: usize>(kind: RuleKind, children: [NamedNode; N]) -> NamedNode {
+    NamedNode {
+        name: None,
+        node: Node::rule(kind, children.into_iter().collect()),
+    }
+}
+
+fn binding_cursors_to_strings(
+    bindings: HashMap<String, Vec<Cursor>>,
+) -> BTreeMap<String, Vec<String>> {
+    bindings
+        .into_iter()
+        .map(|(key, values)| {
+            (
+                key,
+                values
+                    .iter()
+                    .map(|v| v.node().unparse())
+                    .collect::<Vec<_>>(),
+            )
+        })
+        .collect::<BTreeMap<_, _>>()
+}
+
+#[allow(unused_macro_rules)]
+macro_rules! cst_tree {
+    ( @inner [ $($child:expr)* ]) => { [ $($child),* ] };
+
+    ( @inner [ $($child:expr)* ] $kind:ident $text:literal $(, $($rest:tt)*)? ) => {
+        cst_tree!(@inner [ $($child)* token(TokenKind::$kind, $text) ] $($($rest)*)?)
+    };
+
+    ( @inner [ $($child:expr)* ] $kind:ident [ $($children:tt)* ] $(, $($rest:tt)*)? ) => {
+        cst_tree!(@inner [ $($child)* rule(RuleKind::$kind, cst_tree!(@inner [] $($children)*)) ] $($($rest)*)?)
+    };
+
+    // Start with a rule
+    ( $kind:ident [ $($children:tt)* ] ) => {
+        rule(RuleKind::$kind, cst_tree!(@inner [] $($children)*))
+    };
+
+}
+
+macro_rules! query_result {
+    ( $( { $( $key:ident : [ $($value:literal),* ] ),* } )* ) => {
+        vec![ $( {
+            #[allow(unused_mut)]
+            let mut bindings = BTreeMap::new();
+            $( bindings.insert( stringify!($key).to_string(), vec![ $( $value.to_string() ),* ]); )*
+            bindings
+        } ),* ]
+    };
+
+}
+
+fn run_query_test(tree: &NamedNode, query: &str, results: Vec<BTreeMap<String, Vec<String>>>) {
+    let cursor = tree.cursor_with_offset(TextIndex::ZERO);
+    let query = vec![Query::parse(query).unwrap()];
+    let mut results = results.into_iter();
+    for (_query_no, bindings) in cursor.query(query) {
+        let bindings = binding_cursors_to_strings(bindings);
+        if let Some(expected_bindings) = results.next() {
+            assert_eq!(bindings, expected_bindings);
+        } else {
+            panic!("Unexpected query result");
+        }
+    }
+    assert!(results.next().is_none(), "Missing query result");
+}
+
+fn common_test_tree() -> NamedNode {
+    cst_tree!(
```

Comment:
thoughts about using Solidity code snippets as input vs `cst_tree!()`?

Reply:
Originally I thought the same thing, but in this crate we have no solidity :)

Reply:
Oh right 😅

Sounds great. thanks for explaining.